subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c

```diff
@@ -0,0 +1,1251 @@
+		LOG_ERR("Received invalid block index %d", tx_block_index);
+		return -EINVAL;
+	}
+
+	/* Release. */
+	r = release_tx_blocks(dev_data, tx_block_index, size, -1);
+	if (r < 0) {
+		return r;
+	}
+
+	/* When releasing bound message, validate RX state of the endpoint. */
+	if (ept_addr != EPT_ADDR_INVALID) {
+		ept = get_ept_and_rx_validate(dev_data, ept_addr);
+		if (ept == NULL) {
+			LOG_ERR("Received invalid endpoint address %d", ept_addr);
+			r = -EINVAL;
+		}
+	}
+
+	return r;
+}
+
+/**
+ * Bound endpoint message received.
+ */
+static int received_bound(struct backend_data *dev_data, size_t rx_block_index,
+			  uint8_t ept_addr)
+{
+	const struct icmsg_with_buf_config *conf = dev_data->conf;
+	size_t size;
+	uint8_t *buffer;
+
+	/* Validate */
+	buffer = buffer_from_index_validate(&conf->rx, rx_block_index, &size, true);
+	if (buffer == NULL) {
+		LOG_ERR("Received invalid block index %d", rx_block_index);
+		return -EINVAL;
+	}
+
+	/* Put message to waiting array. */
+	k_mutex_lock(&dev_data->mutex, K_FOREVER);
+	dev_data->waiting_bound[ept_addr] = rx_block_index;
+	k_mutex_unlock(&dev_data->mutex);
+
+	/* Schedule processing the message. */
+	schedule_ept_bound_process(dev_data);
+
+	return 0;
+}
+
+/**
+ * Callback called by ICMsg that handles message (data or endpoint bound) received
+ * from the remote.
+ *
+ * @param[in] data	Message received from the ICMsg.
+ * @param[in] len	Number of bytes of data.
+ * @param[in] priv	Opaque pointer to device instance.
+ */
+static void received(const void *data, size_t len, void *priv)
+{
+	const struct device *instance = priv;
+	struct backend_data *dev_data = instance->data;
+	const struct icmsg_message *message = (const struct icmsg_message *)data;
+	uint8_t ept_addr;
+	int r = 0;
+
+	/* Allow messages longer than 3 bytes, e.g. for future protocol versions. */
+	if (len < sizeof(struct icmsg_message)) {
+		r = -EINVAL;
+		goto exit;
+	}
+
+	ept_addr = message->ept_addr;
+	if (ept_addr >= NUM_EPT) {
+		r = -EINVAL;
+		goto exit;
+	}
+
+	switch (message->msg_type) {
+	case MSG_RELEASE_DATA:
+		ept_addr = EPT_ADDR_INVALID;
+		/* Fallback to next case. */
+	case MSG_RELEASE_BOUND:
+		r = received_release_data(dev_data, message->block_index, ept_addr);
+		break;
+	case MSG_BOUND:
+		r = received_bound(dev_data, message->block_index, ept_addr);
+		break;
+	case MSG_DATA:
+		r = received_data(dev_data, message->block_index, ept_addr);
+		break;
+	default:
+		/* Silently ignore other messages types. They can be used in future
+		 * protocol version.
+		 */
+		break;
+	}
+
+exit:
+	if (r < 0) {
+		LOG_ERR("Failed to receive, err %d", r);
+	}
+}
+
+/**
+ * Callback called when ICMsg is bound.
+ */
+static void bound(void *priv)
+{
+	const struct device *instance = priv;
+	struct backend_data *dev_data = instance->data;
+
+	/* Set flag that ICMsg is bounded and now, endpoint bounding may start. */
+	atomic_or(&dev_data->flags, FLAG_ICMSG_BOUNDED);
+	schedule_ept_bound_process(dev_data);
```

Comment:
Does it make sense to trigger bound process on no initiator side on ICMsg bound since it will be triggered on every bound message reception?

Reply:
But the user can call bound of endpoint before ICMsg is bounded (or later), that is always scheduled when both are conditions are true.